src/codelabs/null-safety.md

```diff
@@ -0,0 +1,362 @@
+For example, if you have a variable of
+type `String`, it will always contain a string.
+
+If you want a variable of type `String` to accept any string
+or the value `null`, but nothing else,
+give the variable a nullable type by adding a `?` after the
+type name. For example, a variable of type `String?` can contain a string, or
+it can be null.
+
+### Example: Introducing non-nullable types
+
+The variable `a` below is declared as an `int`. Try changing the value in the
+assignment to 3 or 145. Anything but null!
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/non_nullable_types.dart" replace="/145/null/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: Nullable types
+
+What if you need a variable that *can* hold a null value?  Try changing the
+type of `a` so that `a` can be either null or an int:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/nullable_types.dart" replace="/int\?/int/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: More nullable types!
+
+Type parameters for generics can also be nullable or non-nullable. Try using
+question marks to correct the type declarations of `aNullableListOfStrings` and
+`aListOfNullableStrings`:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/more_nullable_types.dart" replace="/String\?/String/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  List<String> aListofStrings = ['one', 'two', 'three'];
+  List<String> aNullableListOfStrings = [];
+  List<String> aListofNullableStrings = ['one', null, 'three'];
+
+  print('aListofStrings is $aListofStrings.');
+  print('aNullableListOfStrings is $aNullableListOfStrings.');
+  print('aListofNullableStrings is $aListofNullableStrings.');
+}
+```
+
+### Example: Definite assignment
+
+Dart's type system is crafty enough to track where variables are assigned and
+where their values are read, and to verify that non-nullable fields are given
+values before any code tries to read from them. This process is called
+"definite assignment."
+
+Try uncommenting the if-else statement in the code below, and
+watch the analyzer errors disappear:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/definite_assignment.dart" replace="/if/\/\/if/g; /\ \ text\ =/\/\/  text =/g; /\ \ \}/  \/\/}/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  String text;
+
+  //if (DateTime.now().hour < 12) {
+  //  text = "It's morning! Let's make aloo paratha!";
+  //} else {
+  //  text = "It's afternoon! Let's make biryani!";
+  //}
+
+  print(text);
+  print(text.length);
+}
+```
+
+## New null-aware operators: ? and !
+
+Null safety introduces two operators, ? and !. Here's how you use them:
```

Comment:
Rewrite this intro... somehow... Maybe something about dealing with the possibility of null values? (So this isn't about new operators, it's about syntax you're likely to see that we haven't already discussed.)

Reply:
done, PTAL